### PR TITLE
feat: Add deposit cap (SC-1131)

### PR DESCRIPTION
**Pull request by hacker-DOM**

### CODE REVIEW

File: src/ISparkVault.sol [hunk 1]
```diff
@@ -39,6 +39,13 @@ interface ISparkVault is IERC20Permit, IERC4626, IAccessControlEnumerable {
      */
     event Referral(uint16 indexed referral, address indexed owner, uint256 assets, uint256 shares);
 
+    /**
+     * @notice Emitted when the maximum deposit cap is updated.
+     * @param  oldCap The previous maximum deposit cap [wei]
+     * @param  newCap The new maximum deposit cap [wei]
```
Comment (lucas-manuel) [inline src/ISparkVault.sol]: Can't say wei here cause its in underlying asset's decimals, can just remove

File: src/ISparkVault.sol [hunk 2]
```diff
@@ -95,14 +102,22 @@ interface ISparkVault is IERC20Permit, IERC4626, IAccessControlEnumerable {
      */
     function rho() external view returns (uint64);
 
+    /**
+     * @notice Sets the deposit cap for the vault.
+     * @dev    This function can only be called by accounts with DEFAULT_ADMIN_ROLE.
+               Deposits (and mints) are disabled if it would put totalAssets() above this value
+     * @param  newCap The new deposit cap value [wei]
```
Comment (lucas-manuel) [inline src/ISparkVault.sol]: Rm wei

File: src/SparkVault.sol [hunk 3]
```diff
@@ -67,6 +67,8 @@ contract SparkVault is AccessControlEnumerableUpgradeable, UUPSUpgradeable, ISpa
     uint256 public minVsr; // The minimum Vault Savings Rate [ray]
     uint256 public maxVsr; // The maximum Vault Savings Rate [ray]
 
+    // Deposits (and mints) are disabled if it would put totalAssets() above this value [wad/wei]
```
Comment (lucas-manuel) [inline src/SparkVault.sol]: Can rm comment

File: test/Authorization.t.sol [hunk 4]
```diff
@@ -168,6 +168,40 @@ contract SparkVaultRevokeRoleSuccessTests is SparkVaultGrantRoleSuccessTests {
...
+
+contract SparkVaultSetDepositCapSuccessTests is SparkVaultTestBase {
+
+    event DepositCapSet(uint256 oldCap, uint256 newCap);
+
+    function test_setDepositCap() public {
+        assertEq(vault.depositCap(), 1_000_000e6);
+
+        vm.startPrank(admin);
+        vm.expectEmit(true, true, true, true, address(vault));
```
Comment (lucas-manuel) [inline test/Authorization.t.sol]: Can just be vm.expectEmit(address(vault)) here for both

File: test/Authorization.t.sol [hunk 5]
```diff
@@ -168,6 +168,40 @@ contract SparkVaultRevokeRoleSuccessTests is SparkVaultGrantRoleSuccessTests {
...
+        emit DepositCapSet(1_000_000e6, 2_000_000e6);
+        vault.setDepositCap(2_000_000e6);
+        assertEq(vault.depositCap(), 2_000_000e6);
```
Comment (lucas-manuel) [inline test/Authorization.t.sol:196]: Put this on new line below group of lines

File: test/Authorization.t.sol [hunk 6]
```diff
@@ -168,6 +168,40 @@ contract SparkVaultRevokeRoleSuccessTests is SparkVaultGrantRoleSuccessTests {
...
+
+        vm.expectEmit(true, true, true, true, address(vault));
+        emit DepositCapSet(2_000_000e6, type(uint256).max);
+        vault.setDepositCap(type(uint256).max);
+        assertEq(vault.depositCap(), type(uint256).max);
```
Comment (lucas-manuel) [inline test/Authorization.t.sol:202]: Put this on new line below group of lines

Comment (lucas-manuel) [reply test/Authorization.t.sol:202]: Add one more to set it to zero also

File: test/ValueAccrual.t.sol [hunk 7]
```diff
@@ -92,6 +92,9 @@ contract ValueAccrualE2ETest is SparkVaultTestBase {
 
         // Step 5: User 2 deposits 1M assets
 
+        vm.prank(admin);
```
Comment (lucas-manuel) [inline test/ValueAccrual.t.sol]: This should be in the setup

File: test/ERC4626.t.sol [hunk 8]
```diff
@@ -18,6 +18,10 @@ contract SparkVaultERC4626StandardTest is ERC4626Test, SparkVaultTestBase {
     function setUp() public virtual override(ERC4626Test, SparkVaultTestBase) {
         super.setUp();
 
+        // For the purposes of this test, set unlimited deposit cap
+        vm.prank(admin);
+        vault.setDepositCap(type(uint256).max);
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol:19]: Need boundary condition failure mode testing for both mint and deposit.

Also should have a test where we set it to zero after an existing deposit and then withdraw (just to demonstrate withdraw-only functionality)

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: src/ISparkVault.sol (hunk 1 above)
Comment (hacker-DOM) [reply src/ISparkVault.sol]: Good point. There are a lot of other mentions, I'll change it to [asset units] (or can remove it if you want).

Comment (hacker-DOM) [reply src/ISparkVault.sol]: ✅

File: src/ISparkVault.sol (hunk 2 above)
Comment (hacker-DOM) [reply src/ISparkVault.sol]: ✅

File: src/SparkVault.sol (hunk 3 above)
Comment (hacker-DOM) [reply src/SparkVault.sol]: ✅

File: test/Authorization.t.sol (hunk 4 above)
Comment (hacker-DOM) [reply test/Authorization.t.sol]: ✅

File: test/Authorization.t.sol (hunk 5 above)
Comment (hacker-DOM) [reply test/Authorization.t.sol:196]: ✅

File: test/Authorization.t.sol (hunk 6 above)
Comment (hacker-DOM) [reply test/Authorization.t.sol:202]: ✅

File: test/ValueAccrual.t.sol (hunk 7 above)
Comment (hacker-DOM) [reply test/ValueAccrual.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 8 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol:19]: ✅

File: src/SparkVault.sol [hunk 9]
```diff
@@ -315,12 +321,44 @@ contract SparkVault is AccessControlEnumerableUpgradeable, UUPSUpgradeable, ISpa
         return assets * RAY / nowChi();
     }
 
-    function maxDeposit(address) external pure returns (uint256) {
-        return type(uint256).max;
-    }
+    function maxDeposit(address) external view returns (uint256) {
```
Comment (lucas-manuel) [inline src/SparkVault.sol:324]:
```solidity

    function maxDeposit(address) external view returns (uint256) {
        uint256 totalAssets_ = totalAssets();
        uint256 depositCap_  = depositCap;
        return depositCap_ <= totalAssets_ ? 0 : depositCap_ - totalAssets_;
    }

    function maxMint(address) external view returns (uint256) {
        uint256 depositCap_ = depositCap;

        // NOTE: Prevents overflow on (depositCap_ - totalAssets()) * RAY below, values above
        //       type(uint256).max / RAY are considered "infinite".
        if (depositCap_ > type(uint256).max / RAY) return type(uint256).max;

        return (depositCap_ - totalAssets()) * RAY / nowChi();
    }
```

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: src/SparkVault.sol (hunk 9 above)
Comment (hacker-DOM) [reply src/SparkVault.sol:324]: ✅

File: test/ERC4626.t.sol [hunk 10]
```diff
@@ -159,6 +166,74 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+    function test_maxDeposit_depositCapLeTotalAssets() public {
+        // Deposit cap is currently 2_100_000e6 and total assets is slightly above 1_000_000e6
+        vm.prank(admin);
+        vault.setDepositCap(1_000_000e6);
+
+        uint256 maxDepositUser1 = vault.maxDeposit(user1);
+        uint256 maxDepositUser2 = vault.maxDeposit(user2);
+
+        assertEq(maxDepositUser1, 0);
+        assertEq(maxDepositUser2, 0);
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: We should actually change this to:

1. Set cap to totalAssets() - 1, show its 0
2. Set cap to totalAssets(), show its 0
3. Set cap to totalAssets() + 1, show its 1

File: test/ERC4626.t.sol [hunk 11]
```diff
@@ -159,6 +166,74 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+    }
+
+    function test_maxDeposit_depositCapGtTotalAssets() public {
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]:
```solidity
    function test_maxDeposit_depositCapGtTotalAssets() public view {
        // Deposit cap is currently 2_100_000e6 and total assets is slightly above 1_000_000e6
        uint256 maxDepositUser1 = vault.maxDeposit(user1);
        uint256 maxDepositUser2 = vault.maxDeposit(user2);

        assertEq(maxDepositUser1, 1_099_892.540218e6);
        assertEq(maxDepositUser1, maxDepositUser2);

        assertEq(maxDepositUser1, vault.totalAssets() - vault.depositCap());
    }
```

File: test/ERC4626.t.sol [hunk 12]
```diff
@@ -159,6 +166,74 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+        assertEq(maxDepositUser1, 1_099_892.540218e6);
+        assertEq(maxDepositUser2, 1_099_892.540218e6);
+    }
+
+    function test_maxMint_depositCapLeTotalAssets() public {
+        // Deposit cap is currently 2_100_000e6 and total assets is slightly above 1_000_000e6
+        vm.prank(admin);
+        vault.setDepositCap(1_000_000e6);
+
+        uint256 maxMintUser1 = vault.maxMint(user1);
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: Same as maxDeposit, should be boundary based

File: test/ERC4626.t.sol [hunk 13]
```diff
@@ -159,6 +166,74 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+        // => 1e77 / RAY < uint_max / RAY < 1e78 / RAY
+        // => 1e50       < uint_max / RAY < 1e51
+        // So 1e51 should still trigger it (return uint_max)
+        vault.setDepositCap(1e51);
+
+        maxMintUser1 = vault.maxMint(user1);
+        maxMintUser2 = vault.maxMint(user2);
+
+        assertEq(maxMintUser1, type(uint256).max);
+        assertEq(maxMintUser2, type(uint256).max);
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: Lets try setting it to 1e51 - 1 or something to get a different result

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: test/ERC4626.t.sol (hunk 10 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 11 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol:181]: ✅

Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅ I think the last expression should be `vault.depositCap() - vault.totalAssets()` but yes

File: test/ERC4626.t.sol (hunk 12 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 13 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol [hunk 14]
```diff
@@ -159,6 +166,116 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+
+        vm.prank(admin);
+        vault.setDepositCap(1_000_107.459782e6 + 1);
+        maxDepositUser1 = vault.maxDeposit(user1);
+        maxDepositUser2 = vault.maxDeposit(user2);
+        assertEq(maxDepositUser1, 1);
+        assertEq(maxDepositUser2, maxDepositUser1);
+    }
+
+    function test_maxDeposit_depositCapGtTotalAssets() public {
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: view

File: test/ERC4626.t.sol [hunk 15]
```diff
@@ -159,6 +166,116 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+        vm.prank(admin);
+        vault.setDepositCap(1_000_107.459782e6 + 2);
+        maxMintUser1 = vault.maxMint(user1);
+        maxMintUser2 = vault.maxMint(user2);
+        // After increasing depositCap (assets) by 2, shares starts to reflect that
+        assertEq(maxMintUser1, 1);
+        assertEq(maxMintUser2, maxMintUser1);
+    }
+
+    function test_maxMint_depositCapGtTotalAssets() public {
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: view

File: test/ERC4626.t.sol [hunk 16]
```diff
@@ -159,6 +166,116 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+        // Deposit cap is currently 2_100_000e6 and total assets is slightly above 1_000_000e6
+        uint256 maxMintUser1 = vault.maxMint(user1);
+        uint256 maxMintUser2 = vault.maxMint(user2);
+
+        assertEq(maxMintUser1, 1_099_774.358705e6);
+        assertEq(maxMintUser2, 1_099_774.358705e6);
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: Can just inline these

File: test/ERC4626.t.sol [hunk 17]
```diff
@@ -159,6 +166,116 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+    }
+
+    function test_maxMint_extremelyLargeDepositCap() public {
+        // Deposit cap is currently 2_100_000e6 and total assets is slightly above 1_000_000e6
+        vm.prank(admin);
+        vault.setDepositCap(type(uint256).max);
+        uint256 maxMintUser1 = vault.maxMint(user1);
+        uint256 maxMintUser2 = vault.maxMint(user2);
+        assertEq(maxMintUser1, type(uint256).max);
+        assertEq(maxMintUser2, maxMintUser1);
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: Inline all of these

File: test/ERC4626.t.sol [hunk 18]
```diff
@@ -159,6 +166,116 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+        assertEq(maxDepositUser1, vault.depositCap() - vault.totalAssets());
+    }
+
+    function test_maxMint_depositCapLeTotalAssets() public {
+        // Deposit cap is currently 2_100_000e6 and total assets is slightly above 1_000_000e6
+        assertEq(vault.totalAssets(), 1_000_107.459782e6);
+
+        vm.prank(admin);
+        vault.setDepositCap(1_000_107.459782e6 - 1);
+        uint256 maxMintUser1 = vault.maxMint(user1);
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: Inline

File: test/ERC4626.t.sol [hunk 19]
```diff
@@ -159,6 +166,116 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
     }
 
+    function test_maxDeposit_depositCapLeTotalAssets() public {
+        // Deposit cap is currently 2_100_000e6 and total assets is slightly above 1_000_000e6
+        vm.prank(admin);
+        vault.setDepositCap(1_000_107.459782e6 - 1);
+        uint256 maxDepositUser1 = vault.maxDeposit(user1);
+        uint256 maxDepositUser2 = vault.maxDeposit(user2);
+        assertEq(maxDepositUser1, 0);
+        assertEq(maxDepositUser2, maxDepositUser1);
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: assertEq's should always be newlined

Comment (lucas-manuel) [reply test/ERC4626.t.sol]: Also can just remove all the varirables and inline the assertions

File: test/ERC4626.t.sol [hunk 20]
```diff
@@ -159,6 +166,116 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+    }
+
+    function test_maxDeposit_depositCapGtTotalAssets() public {
+        // Deposit cap is currently 2_100_000e6 and total assets is slightly above 1_000_000e6
+        uint256 maxDepositUser1 = vault.maxDeposit(user1);
+        uint256 maxDepositUser2 = vault.maxDeposit(user2);
+
+        assertEq(maxDepositUser1, 1_099_892.540218e6);
+        assertEq(maxDepositUser2, maxDepositUser1);
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: Inline

File: test/ERC4626.t.sol [hunk 21]
```diff
@@ -159,6 +166,116 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+
+        assertEq(maxDepositUser1, vault.depositCap() - vault.totalAssets());
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: Inline

File: test/ERC4626.t.sol [hunk 22]
```diff
@@ -159,6 +166,116 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+        uint256 maxMintUser2 = vault.maxMint(user2);
+        assertEq(maxMintUser1, 0);
+        assertEq(maxMintUser2, maxMintUser1);
+
+        vm.prank(admin);
+        vault.setDepositCap(1_000_107.459782e6);
+        maxMintUser1 = vault.maxMint(user1);
+        maxMintUser2 = vault.maxMint(user2);
+        assertEq(maxMintUser1, 0);
+        assertEq(maxMintUser2, maxMintUser1);
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: Same here, newline and inline

File: test/ERC4626.t.sol [hunk 23]
```diff
@@ -159,6 +166,116 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+        // It returns uint_max when depositCap > uint_max / RAY
+        //    1e77       < uint_max       < 1e78
+        // => 1e77 / RAY < uint_max / RAY < 1e78 / RAY
+        // => 1e50       < uint_max / RAY < 1e51
+        // So 1e51 should still trigger it (return uint_max)
+        vault.setDepositCap(1e51);
+        maxMintUser1 = vault.maxMint(user1);
+        maxMintUser2 = vault.maxMint(user2);
+        assertEq(maxMintUser1, type(uint256).max);
+        assertEq(maxMintUser2, maxMintUser1);
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: Newline and inline for all of these

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: test/ERC4626.t.sol (hunk 14 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol [hunk 24]
```diff
@@ -159,6 +166,116 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+        // Deposit cap is currently 2_100_000e6 and total assets is slightly above 1_000_000e6
+        vm.prank(admin);
+        vault.setDepositCap(1_000_107.459782e6 - 1);
+        uint256 maxDepositUser1 = vault.maxDeposit(user1);
+        uint256 maxDepositUser2 = vault.maxDeposit(user2);
+        assertEq(maxDepositUser1, 0);
+        assertEq(maxDepositUser2, maxDepositUser1);
+
+        vm.prank(admin);
+        vault.setDepositCap(1_000_107.459782e6);
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: Use totalAssets()

File: test/ERC4626.t.sol [hunk 25]
```diff
@@ -159,6 +166,116 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+
+        assertEq(maxDepositUser1, vault.depositCap() - vault.totalAssets());
+    }
+
+    function test_maxMint_depositCapLeTotalAssets() public {
+        // Deposit cap is currently 2_100_000e6 and total assets is slightly above 1_000_000e6
+        assertEq(vault.totalAssets(), 1_000_107.459782e6);
+
+        vm.prank(admin);
+        vault.setDepositCap(1_000_107.459782e6 - 1);
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: Same changes as in deposit, use totalAssets - 1, use 0 and 1 in assertions, and put everything on new lines

File: test/ERC4626.t.sol [hunk 26]
```diff
@@ -159,6 +166,116 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+        assertEq(maxMintUser1, 1);
+        assertEq(maxMintUser2, maxMintUser1);
+    }
+
+    function test_maxMint_depositCapGtTotalAssets() public {
+        // Deposit cap is currently 2_100_000e6 and total assets is slightly above 1_000_000e6
+        uint256 maxMintUser1 = vault.maxMint(user1);
+        uint256 maxMintUser2 = vault.maxMint(user2);
+
+        assertEq(maxMintUser1, 1_099_774.358705e6);
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: Should derive this number, like (vault.depositCap - vault.totalAssets) * 1e27 / vault.nowChi()

File: test/ERC4626.t.sol (hunk 17 above)
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: Same here, inline, new lines and hard number assertions

File: test/ERC4626.t.sol [hunk 27]
```diff
@@ -566,6 +717,37 @@ contract SparkVaultWithdrawSuccessTests is SparkVaultTestBase {
...
+        assertEq(vault.balanceOf(user1),          1_000_000e6);
+        assertEq(vault.assetsOf(user1),           assets);
+        assertEq(vault.totalAssets(),             assets);
+        assertEq(asset.balanceOf(address(vault)), assets);
+        assertEq(asset.balanceOf(user1),          0);
+
+        // Set deposit cap to 0
+        vm.prank(admin);
+        vault.setDepositCap(0);
+
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol:718]: Lets just add a quick line to show they can't deposit but can withdraw, just try depositing 1 unit here.

File: test/ERC4626.t.sol [hunk 28]
```diff
@@ -566,6 +717,37 @@ contract SparkVaultWithdrawSuccessTests is SparkVaultTestBase {
         assertEq(asset.balanceOf(randomUser),     assets);
     }
 
+    function test_withdraw_depositCapExceeded() public {
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: test_withdraw_succeedsWithDepositCapZero

File: test/ERC4626.t.sol [hunk 29]
```diff
@@ -726,4 +908,38 @@ contract SparkVaultRedeemSuccessTests is SparkVaultTestBase {
         assertEq(asset.balanceOf(randomUser),     assets);
     }
 
+    function test_redeem_depositCapExceeded() public {
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: test_redeem_succeedsWithDepositCapZero

File: test/ERC4626.t.sol [hunk 30]
```diff
@@ -726,4 +908,38 @@ contract SparkVaultRedeemSuccessTests is SparkVaultTestBase {
...
+        assertEq(vault.balanceOf(user1),          1_000_000e6);
+        assertEq(vault.assetsOf(user1),           assets);
+        assertEq(vault.totalAssets(),             assets);
+        assertEq(asset.balanceOf(address(vault)), assets);
+        assertEq(asset.balanceOf(user1),          0);
+
+        // Set deposit cap to 0
+        vm.prank(admin);
+        vault.setDepositCap(0);
+
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol:918]: Add revert case for mint(1)

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: test/ERC4626.t.sol (hunk 15 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 16 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 17 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 18 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 19 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 20 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 21 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 22 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 23 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 24 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 25 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 26 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 27 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol:718]: ✅

File: test/ERC4626.t.sol (hunk 28 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 29 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 30 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol:918]: ✅

File: test/ERC4626.t.sol [hunk 31]
```diff
@@ -159,6 +162,103 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
         vault.previewWithdraw(assets);
     }
 
+    function test_maxDeposit_depositCapLeTotalAssets() public {
+        // Deposit cap is currently 2_100_000e6 and total assets is slightly above 1_000_000e6
+        uint256 totalAssets = vault.totalAssets();
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: This is actually already in storage from setup so don't need to cache to a variable

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: test/ERC4626.t.sol [hunk 32]
```diff
@@ -159,6 +162,103 @@ contract SparkVaultERC4626Test is SparkVaultTestBase {
...
+
+        vm.prank(admin);
+        vault.setDepositCap(totalAssets + 2);
+
+        // After increasing depositCap (assets) by 2, shares starts to reflect that
+        assertEq(vault.maxMint(user1), 1);
+        assertEq(vault.maxMint(user2), 1);
+    }
+
+    function test_maxMint_depositCapGtTotalAssets() public {
```
Comment (lucas-manuel) [inline test/ERC4626.t.sol]: public view

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

File: test/ERC4626.t.sol (hunk 31 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅

File: test/ERC4626.t.sol (hunk 32 above)
Comment (hacker-DOM) [reply test/ERC4626.t.sol]: ✅
